### PR TITLE
An alternative to #249

### PR DESCRIPTION
**Pull request by skirpichev**

### CODE REVIEW

**Comment by vinklein** [comment]
Nice.
It make sense to have HAS_***_CONVERSION tested in last to avoid `PyObject_HasAttrString` calls if possible.
All my tests pass with this branch.
Looks good to me.

**Comment by skirpichev** [comment]
Ok, #249 is closed in favor of this work.

**Comment by casevh** [comment]
Thank you for tracking this bug down.

**Comment by skirpichev** [comment]
You are welcome!

BTW, @casevh, is there any plans for b2?

File: src/gmpy2_convert.h [hunk 1]
```diff
@@ -46,20 +46,27 @@ extern "C" {
...
+                        HAS_MPZ_CONVERSION(x))
 #endif
 
-#define IS_FRACTION(x) (!strcmp(Py_TYPE(x)->tp_name, "Fraction"))
-
-#define IS_RATIONAL_ONLY(x) (MPQ_Check(x) || IS_FRACTION(x) || HAS_MPQ_CONVERSION(x))
-#define IS_RATIONAL(x) (IS_INTEGER(x) || IS_RATIONAL_ONLY(x))
-
 #define IS_REAL_ONLY(x) (MPFR_Check(x) || PyFloat_Check(x) || HAS_STRICT_MPFR_CONVERSION(x))
 #define IS_REAL(x) (IS_RATIONAL(x) || IS_REAL_ONLY(x))
```
Comment (isuruf) [inline src/gmpy2_convert.h]: Shouldn't these be fixed as well?

**Comment by skirpichev** [comment]
On Tue, Jul 09, 2019 at 07:58:54AM -0700, Isuru Fernando wrote:
>   #define IS_REAL_ONLY(x) (MPFR_Check(x) || PyFloat_Check(x) || HAS_STRICT_MPFR_CONVERSION(x))
>   #define IS_REAL(x) (IS_RATIONAL(x) || IS_REAL_ONLY(x))
> 
>    Shouldn't these be fixed as well?

Yeah, this PR was a causal fix, related to the diofant test
suite.  Diofant (or SymPy, AFAIK), doesn't use gmpy's floats.

But I cat bet - these macroses have same issue.

**Pull request merged**